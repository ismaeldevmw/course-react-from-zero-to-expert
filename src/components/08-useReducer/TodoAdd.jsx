import React, { useRef } from 'react'

export const TodoAdd = ({ handleAddTask }) => {

    const inputRef = useRef();

    const handleSubmit = (e) => {

        e.preventDefault();

        if (inputRef.current.value.trim().length <= 1) return;

        const newTask = {
            id: new Date().getTime(),
            description: inputRef.current.value,
            done: false
        };

        handleAddTask( newTask );
        inputRef.current.value = '';

    }

    return (
        <>
            <h4>Add task</h4>
            <hr/>

            <form onSubmit={ handleSubmit } className="form-inline">
                <div className="form-group mx-sm-3 mb-2">
                    <input 
                        type="text"
                        name="description"
                        placeholder="Add a task to do..."
                        autoComplete="off"
                        className="form-control form-control-sm"
                        ref={ inputRef }
                    />
                </div>
                <button
                    type="submit" 
                    className="btn btn-sm btn-outline-primary mb-2"
                >
                    Add
                </button>
            </form>
        </>
    )
}
