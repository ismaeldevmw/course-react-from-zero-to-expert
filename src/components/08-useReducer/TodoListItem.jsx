import React from 'react'

export const TodoListItem = ({ task, index, handleToggle, handleDelete }) => {
    return (
        <li
            key={ task.id }
            className="list-group-item d-flex justify-content-between align-items-center"
        >
            <span
                className={ `${ task.done && 'complete' }` } 
                onClick={ () => handleToggle( task.id ) } 
            >
                { index + 1}. { task.description }
            </span>
            <button onClick={ () => handleDelete(task.id) } className="btn btn-sm btn-danger">Delete</button>
        </li>
    )
}
