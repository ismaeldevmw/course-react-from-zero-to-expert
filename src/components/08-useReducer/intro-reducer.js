const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false
}];

const todoReducer = ( state = initialState, action ) => {
    
    if ( action?.type === 'add' ) {
        return [ ...state, action.payload ];
    }

    return state;
}

let todos = todoReducer();


const newTodo = {
    id: 2,
    todo: 'But eggs',
    done: false
}


const addElementAction = {
    type: 'add',
    payload: newTodo
}


todos = todoReducer( todos, addElementAction );

console.log(todos);