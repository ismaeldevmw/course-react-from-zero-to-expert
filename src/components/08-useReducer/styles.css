body {
    padding: 3rem;
}

.complete {
    text-decoration: line-through;
}