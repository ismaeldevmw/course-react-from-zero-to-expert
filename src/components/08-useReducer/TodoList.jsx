import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ tasks = [], handleToggle, handleDelete }) => {
    return (
        <ul className="list-group">
            {
                tasks.map( (task, i) => (
                    <TodoListItem
                        key={ task.id }
                        task={ task } 
                        index={ i } 
                        handleToggle={ handleToggle }
                        handleDelete={ handleDelete } 
                    />
                ))
            }
        </ul>
    )
}
