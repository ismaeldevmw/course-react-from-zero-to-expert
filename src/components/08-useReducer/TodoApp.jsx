import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.css';

const init = () => {

    return JSON.parse(localStorage.getItem('tasks')) || []
    // return [{
    //     id: new Date().getTime(),
    //     description: 'Learn React',
    //     done: false
    // }];

}

export const TodoApp = () => {

    const [ taskToDo, dispatch ] = useReducer(todoReducer, [], init);

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    useEffect( () => {
        localStorage.setItem('tasks', JSON.stringify( taskToDo ));
    }, [ taskToDo ])

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;
        
        const newTask = {
            id: new Date().getTime(),
            description: description,
            done: false
        };

        const action = {
            type: 'add',
            payload: newTask
        }

        dispatch( action);
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({ taskToDo.length })</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                    <ul className="list-group">
                        {
                            taskToDo.map( (task, i) => (
                                <li
                                    key={ task.id }
                                    className="list-group-item d-flex justify-content-between align-items-center"
                                >
                                    { i + 1}. { task.description }
                                    <button className="btn btn-sm btn-danger">Delete</button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Add task</h4>
                    <hr/>

                    <form onSubmit={ handleSubmit } className="form-inline">
                        <div className="form-group mx-sm-3 mb-2">
                            <input 
                                type="text"
                                name="description"
                                placeholder="Add a task to do..."
                                autoComplete="off"
                                className="form-control form-control-sm"
                                value={ description }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <button
                            type="submit" 
                            className="btn btn-sm btn-outline-primary mb-2"
                        >
                            Add
                        </button>
                    </form>
                </div>
            </div>
        </div>
    )
}
