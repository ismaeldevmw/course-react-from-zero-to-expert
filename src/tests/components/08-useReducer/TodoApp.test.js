import React from 'react';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTasks } from '../../fixtures/demoTasks';
import { act } from '@testing-library/react';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';


describe('Tests in <TodoApp />', () => {
    const wrapper = shallow( <TodoApp /> );

    Storage.prototype.setItem = jest.fn( () => {} )

    test('should display correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should add a TASK', () => {
        
        const wrapper = mount( <TodoApp /> );

        act( () => {
            wrapper.find('TodoAdd').prop('handleAddTask')( demoTasks[0] );
            wrapper.find('TodoAdd').prop('handleAddTask')( demoTasks[1] );
        });
        
        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp (2)');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);

    });

    test('should delete a TASK', () => {
        
        wrapper.find('TodoAdd').prop('handleAddTask')( demoTasks[0] );
        wrapper.find('TodoList').prop('handleDelete')( demoTasks[0].id );

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp (0)');

    });
});