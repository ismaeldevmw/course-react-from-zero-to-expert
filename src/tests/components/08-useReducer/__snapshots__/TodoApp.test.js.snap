// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoApp /> should display correctly 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        tasks={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTask={[Function]}
      />
    </div>
  </div>
</div>
`;
